source/extensions/filters/http/cache/http_cache.h

```diff
@@ -267,6 +274,17 @@ class LookupContext {
   // Get the trailers from the cache. Only called if LookupResult::has_trailers
   // == true.
   virtual void getTrailers(LookupTrailersCallback&& cb) PURE;
+
+  /**
+   * This routine is called prior to an LookupContext being destroyed. LookupContext is responsible
+   * for making sure that any async events are cleaned up in the context of this routine. This
+   * includes timers, network calls, etc. The reason there is an onDestroy() method vs. doing this
+   * type of cleanup in the destructor is due to the deferred deletion model that Envoy uses to
+   * avoid stack unwind complications. LookupContext must not invoke any callbacks after having
+   * onDestroy() invoked.
+   */
+  virtual void onDestroy() PURE;
+  virtual ~LookupContext() = default;
```

Comment:
High level question: The RAII pattern you had before here seemed great. Can't you just explicit reset the pointer in the filter's onDestroy() method and not introduce another onDestroy() here?

Reply:
Switching CacheFilter to an onDestroy-based teardown forced me to reconsider (Insert|Lookup)Context's destruction, and I realized that I was forcing unnecessary subtlety on implementors, and that onDestroy was simpler. With the straight destruction approach, implementers need to make sure that any they don't call any virtual functions asynchronously,  to avoid potentially racing with the destructor. This can be done, but I was inviting mistakes by requiring it. By providing an onDestroy method, I follow existing Envoy patterns, and make it clear that asynchronous activity must be finished before onDestroy returns.

Reply:
Fair enough. Personally, I would still use RAII where we can, but we can revisit this later as needed. I will review the rest of the PR today.

Reply:
Reverted to destroying contexts in CacheFilter::onDestroy.